### ISSUE TITLE
Trying to output zero things crashs terraform

### ISSUE THREAD

**Issue by simesy** [hashicorp/terraform#15300]
I feel like there must be a way to handle output of my aws_security_group_rule when its count=0, but the fact I can't find anything in the googles, and that it crashes Terraform, have decided to report it.

### Terraform Version
Terraform v0.9.8

### Terraform config:
See steps to reproduce

### Panic Output
https://gist.github.com/simesy/c9652b0a6ea7daf9dd85294db1d5ce12

### Expected Behavior
Should not crash when accessing aws_security_group_rule output when count=0.

### Actual Behavior
Crashes

### Steps to Reproduce
1. clone https://github.com/simesy/tf_lemp
2. check out 087483eb7503fe1dd9a9ac03695c6b14ee1b4b08
3. Run `terraform apply -var-file tests/test.tfvar -var remote_access=true`
4. Then `terraform apply -var-file tests/test.tfvar -var remote_access=false`

**Comment by radeksimko** [comment]
Thanks for the crash report.

Clipping out relevant part of crash log:

```
panic: runtime error: index out of range

goroutine 1197 [running]:
github.com/hashicorp/terraform/flatmap.expandArray(0xc420af0000, 0xc420585700, 0x7, 0x9, 0xc420780dc8, 0xc420908d01)
	/opt/gopath/src/github.com/hashicorp/terraform/flatmap/expand.go:110 +0x70c
github.com/hashicorp/terraform/flatmap.Expand(0xc420af0000, 0xc420585700, 0x7, 0x2, 0x2)
	/opt/gopath/src/github.com/hashicorp/terraform/flatmap/expand.go:35 +0x41f
github.com/hashicorp/terraform/terraform.(*Interpolater).interpolateComplexTypeAttribute(0xc420549600, 0xc420585700, 0x7, 0xc420af0000, 0x8e17dc0, 0xc420908f00, 0x9, 0x53ce1a0, 0xc42044aaf0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/interpolate.go:663 +0x206
github.com/hashicorp/terraform/terraform.(*Interpolater).computeResourceVariable(0xc420549600, 0xc4209094f8, 0xc4207614a0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/interpolate.go:468 +0xb67
github.com/hashicorp/terraform/terraform.(*Interpolater).valueResourceVar(0xc420549600, 0xc4209094f8, 0xc4205856e2, 0x25, 0xc4207614a0, 0xc420a142a0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/interpolate.go:251 +0x7c
github.com/hashicorp/terraform/terraform.(*Interpolater).Values(0xc420549600, 0xc4209574f8, 0xc420585770, 0x0, 0xc420909550, 0x1091485)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/interpolate.go:86 +0x699
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).Interpolate(0xc4201721c0, 0xc420761320, 0x0, 0x0, 0xc4208aa000, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:298 +0xe8
github.com/hashicorp/terraform/terraform.(*EvalWriteOutput).Eval(0xc4208942c0, 0x8d8dcc0, 0xc4201721c0, 0x0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_output.go:48 +0x96
github.com/hashicorp/terraform/terraform.EvalRaw(0x8d55940, 0xc4208942c0, 0x8d8dcc0, 0xc4201721c0, 0xc42008a000, 0x2, 0xc420a14240, 0x2b)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc4208942e0, 0x8d8dcc0, 0xc4201721c0, 0x2, 0x2, 0x6149175, 0x4)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:14 +0xc1
github.com/hashicorp/terraform/terraform.EvalRaw(0x8d555c0, 0xc4208942e0, 0x8d8dcc0, 0xc4201721c0, 0x2b, 0x0, 0x0, 0x2b)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc420a14150, 0x8d8dcc0, 0xc4201721c0, 0x2, 0x2, 0x6149175, 0x4)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/hashicorp/terraform/terraform.EvalRaw(0x8d55380, 0xc420a14150, 0x8d8dcc0, 0xc4201721c0, 0x567a8c0, 0xc420bc21f6, 0x53ce1a0, 0xc420bc2270)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.Eval(0x8d55380, 0xc420a14150, 0x8d8dcc0, 0xc4201721c0, 0xc420a14150, 0x8d55380, 0xc420a14150, 0xc4207ceb60)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x5c5cb20, 0xc420ae7760, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:126 +0xd4d
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc42038f0a0, 0x5c5cb20, 0xc420ae7760, 0xc420932280)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:387 +0x392
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:310 +0x9ca
```

Also this looks very similar to https://github.com/hashicorp/terraform/issues/15135

**Comment by simesy** [comment]
I'm coming to the conclusion that this is something about aws_security_group_rule, as I've had instability using it in a few ways.

**Comment by apparentlymart** [comment]
Hi @simesy! I just merged #15390, which should fix this. The fix should be included in the next release.

Thanks for reporting it, and sorry for the crash.

**Comment by simesy** [comment]
Thanks so much @apparentlymart
